test/functional/legacy/076_completefunc_spec.lua

```diff
@@ -0,0 +1,66 @@
+-- Tests for completefunc/omnifunc.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, expect = helpers.clear, helpers.expect
+
+describe('completefunc', function()
+  setup(clear)
+
+  it('is working', function()
+    insert([=[
+      +++
+      one
+      two
+      three]=])
+
+    -- Test that nothing happens if the 'completefunc' opens.
+    -- A new window (no completion, no crash).
+    source([=[
+      function! DummyCompleteOne(findstart, base)
+	if a:findstart
+	  return 0
+	else
+	  wincmd n
+	  return ['onedef', 'oneDEF']
+	endif
+      endfunction
+      setlocal completefunc=DummyCompleteOne
+      /^one
+    ]=])
+    feed('A<C-X><C-U><C-N><esc>:q!<cr>')
+    source([=[
+      function! DummyCompleteTwo(findstart, base)
+	if a:findstart
+	  wincmd n
+	  return 0
+	else
+	  return ['twodef', 'twoDEF']
+	endif
+      endfunction
+      setlocal completefunc=DummyCompleteTwo
+      /^two
+    ]=])
+    feed('A<C-X><C-U><C-N><esc>:q!<cr>')
+    -- Test that 'completefunc' works when it's OK.
+    source([=[
```

Comment:
I think using `source` is fine when it isn't overapplied.

In this test here I would extract the script-like parts (function definitions) in a `source` block at the top

```
function! DummyCompleteOne(...)
endfunction

function! DummyCompleteTwo(...)
endfunction

function! DummyCompleteThree(...)
endfunction
```

And then _below_ perform the actual test steps with normal `execute`/`feed` commands.

```
-- Test that nothing happens if the 'completefunc' opens.
-- A new window (no completion, no crash).
execute('setlocal completefunc=DummyCompleteOne')
execute('/^one')
feed('A<C-X><C-U><C-N><esc>:q!<cr>')

-- Test that ...
...
```

I hope this makes my advice about `source` a bit clearer. Perhaps somebody else can say whether they think this approach makes sense.